pkg/controller/reconciler/component_reconciler_actions.go

```diff
@@ -33,28 +35,43 @@ func (a *dynamicWatchAction) run(_ context.Context, rr *types.ReconciliationRequ
 			continue
 		}
 
-		err := a.controller.Watch(
-			source.Kind(a.mgr.GetCache(), w.object),
-			w.eventHandler,
-			w.predicates...,
-		)
+		ok, err := a.shouldWatch(ctx, w, rr)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
 
+		err = a.fn(w.object, w.eventHandler, w.predicates...)
 		if err != nil {
 			return fmt.Errorf("failed to create watcher for %s: %w", w.object.GetObjectKind().GroupVersionKind(), err)
 		}
 
-		DynamicWatchResourcesTotal.WithLabelValues(controllerName).Inc()
 		a.watched[gvk] = struct{}{}
+		DynamicWatchResourcesTotal.WithLabelValues(controllerName).Inc()
 	}
 
 	return nil
 }
 
-func newDynamicWatchAction(mgr ctrl.Manager, controller controller.Controller, watches []watchInput) actions.Fn {
+func (a *dynamicWatchAction) shouldWatch(ctx context.Context, in watchInput, rr *types.ReconciliationRequest) (bool, error) {
+	for pi := range in.dynamicPred {
+		ok, err := in.dynamicPred[pi](ctx, rr)
+		if err != nil {
+			return false, err
```

Comment:
Do I understand correctly: if one of these dynamicPreds returns an error, then that will be returned from the run method, which is the action; and I think that would eventually bubble up to the Reconcile method of the ComponentReconciler, which I guess would prevent subsequent actions from being run?

I wonder if it would be safer to only allow the predicate to return a boolean, to prevent the possibility of it being too fragile/blocking in odd scenarios in production that the predicate author didn't consider?

Reply:
Agree, it was mainly added because I was thinking one would have had the need to interact with the API server, but probably better to avoid over-complicates things at this stage.

Removed and we can add it back if/when it would make sense.